### PR TITLE
handle grad with `stride=0` on GPU MvBackward

### PR DESCRIPTION
**Pull request by kshitij12345**
References : #38315 ,  #29984

cuBlas expects strides to be greater than 0.
Cloning the `grad` allocates a new vector with
non-zero strides.

For CPU, we don't clone and allocate a new vector
as CPU implementation works with stride=0.

### CODE REVIEW

**Comment by kshitij12345** [comment]
https://github.com/pytorch/pytorch/blob/e3584f8d7e2713f7217cfc44747d169e81d6071a/aten/src/ATen/native/cuda/Blas.cu#L7-L11

More comprehensive solution probably would be to have something like
```cpp
if (vec_stride == 0) {
    auto vec_clone = vec.clone();
    vec_stride = vec_clone.stride(0);
}
```
in the above file.


But not sure about the coverage of that change.

**Comment by ngimel** [comment]
Indeed, a fix in pytorch/aten/src/ATen/native/cuda/Blas.cu would be preferable. We are already making a copy of the matrix if none of the matrix strides is 1, we can do the same for vector if vector stride is 0.

File: tools/autograd/derivatives.yaml
Comment (albanD) [inline tools/autograd/derivatives.yaml:626]: Note that the semantic of `.clone()` has changed, it now preserves the layout of the input.
If you just want a contiguous Tensor, you should use `grad.contiguous()` (that will avoid the copy in the case where it already is contiguous as extra benefit).

Comment (kshitij12345) [reply tools/autograd/derivatives.yaml]: Thanks for the information. Have removed `clone` in favour of `contiguous`.

**Review by albanD — APPROVED** [review]
Looks good.
I would be in favor of merging this as it is a small change and fix the original issue.
And leave the more involved fix of BLAS binding to a latter PR.
Are you ok with that @ngimel ?

**Comment by ngimel** [comment]
I'd strongly prefer a fix on the blas side, it is pretty much as simple, and would fix cases where users call .mv on expanded vectors in the forward. Just add 
```
if (vec_stride == 0) vec = vec.contiguous();
```
here
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/Blas.cu#L9

**Comment by albanD** [comment]
Sounds good, waiting on the update to merge.

**Comment by kshitij12345** [comment]
@albanD @ngimel PTAL:)

File: aten/src/ATen/native/cuda/Blas.cu
Comment (ngimel) [inline aten/src/ATen/native/cuda/Blas.cu:9]: You are unconditionally calling vec.contiguous(), whereas you only need to do it if vec_stride of the original vector is equal to 0.

**Review by ngimel — CHANGES_REQUESTED** [review]
Please don't call vec.contiguous() unconditionally. 
Can you also try enabling test_blas_nan_out in test_torch for cuda? It should work when 0-stride for vector is supported.

File: aten/src/ATen/native/cuda/Blas.cu
Comment (kshitij12345) [reply aten/src/ATen/native/cuda/Blas.cu]: Sure. Have updated.

I actually assumed that the unconditional `contiguous` wouldn't be an overhead for already contiguous `vec` and that it would be similar to checking `vec_stride == 0` and the branching on it.

Comment (albanD) [reply aten/src/ATen/native/cuda/Blas.cu]: It won't, but there are many non-contiguous Tensors that don't have stride=0 and that are supported.

File: test/test_autograd.py
Comment (albanD) [inline test/test_autograd.py:6158]: Is that really a Tensor of stride 0?
I think you want: `torch.randn(1, device=device).expand(2).requires_grad_(True)`.
The expand will make the stride be 0.

File: aten/src/ATen/native/cuda/Blas.cu
Comment (kshitij12345) [reply aten/src/ATen/native/cuda/Blas.cu]: Oh, I understand now. Thank You :).

File: test/test_autograd.py
Comment (kshitij12345) [reply test/test_autograd.py]: Oops. 
Fixed. Thanks.

**Review by albanD — APPROVED** [review]
LGTM
Can you rebase on top of master to make sure the latest test pass (there have been some updates to self.assertEqual()) please?
Will merge once @ngimel approves.

**Comment by kshitij12345** [comment]
Sure. Will do that.

**Comment by ngimel** [comment]
Please also try enabling `test_blas_nan_out` that was previously disabled on CUDA because zero strides were not supported.

**Review by ngimel — APPROVED** [review]

**Comment by kshitij12345** [comment]
> Please also try enabling test_blas_nan_out that was previously disabled on CUDA because zero strides were not supported.

Have removed the `onlyCPU` decorator.
https://github.com/pytorch/pytorch/blob/bf83966541e20a780ed69ebf6fca12478fa8f1f6/test/test_torch.py#L13012-L13021

File: test/test_autograd.py
Comment (albanD) [inline test/test_autograd.py:6157]: You will need to change these to `dtype=torch.double` for the gradcheck to be happy.
It is because all the default threshold are set assuming double precision.
This should fix the CI :)

**Comment by kshitij12345** [comment]
I had actually checked the `dtype` for `mat and vec` and both were `torch.float64`.

I guess the issue is something else as it is failing on `cpu`, which was already supported.

```
May 15 16:30:32 ======================================================================
May 15 16:30:32 ERROR [0.003s]: test_mv_grad_stride_0_cpu (__main__.TestAutogradDeviceTypeCPU)
May 15 16:30:32 ----------------------------------------------------------------------
May 15 16:30:32 Traceback (most recent call last):
May 15 16:30:32   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 210, in instantiated_test
May 15 16:30:32     return test(self, device_arg)
May 15 16:30:32   File "test_autograd.py", line 6163, in test_mv_grad_stride_0
May 15 16:30:32     gradcheck(fn, (vec))
May 15 16:30:32   File "/opt/conda/lib/python3.6/site-packages/torch/autograd/gradcheck.py", line 306, in gradcheck
May 15 16:30:32     'numerical:%s\nanalytical:%s\n' % (i, j, n, a))
May 15 16:30:32   File "/opt/conda/lib/python3.6/site-packages/torch/autograd/gradcheck.py", line 252, in fail_test
May 15 16:30:32     raise RuntimeError(msg)
May 15 16:30:32 RuntimeError: Jacobian mismatch for output 0 with respect to input 0,
May 15 16:30:32 numerical:tensor([[-0.3463],
May 15 16:30:32         [-0.3463]])
May 15 16:30:32 analytical:tensor([[-0.9654],
May 15 16:30:32         [ 0.6191]])
```

**Comment by ngimel** [comment]
That's right, autograd test is failing on current master (but it passes if vector is contiguous). So this should hot hold this PR, please file an issue about autograd, and skip autograd test on cpu mentioning this issue in the comment.

**Comment by kshitij12345** [comment]
@ngimel Gentle ping:).

**Pull request closed without merging**
